src/setup/node.rs

```diff
@@ -1,193 +1,187 @@
+    pub fn validator_token(mut self, token: String) -> Self {
+        self.conf.validator_token = Some(token);
+        self
+    }
 
-        // TODO: set initial peers/initial actions.
-        self.generate_config_file()?;
+    /// Sets network's id to form an isolated testnet.
+    pub fn network_id(mut self, network_id: u32) -> Self {
+        self.conf.network_id = Some(network_id);
+        self
+    }
 
-        let (stdout, stderr) = match self.config.log_to_stdout {
+    fn start_node(self, log_to_stdout: bool) -> Node {
+        let (stdout, stderr) = match log_to_stdout {
             true => (Stdio::inherit(), Stdio::inherit()),
             false => (Stdio::null(), Stdio::null()),
         };
-
-        let process = Command::new(&self.meta.start_command)
+        let child = Command::new(&self.meta.start_command)
             .current_dir(&self.meta.path)
             .args(&self.meta.start_args)
             .stdin(Stdio::null())
             .stdout(stdout)
             .stderr(stderr)
             .spawn()
             .expect("node failed to start");
-
-        self.wait_for_start().await;
-        self.process = Some(process);
-
-        Ok(())
-    }
-
-    async fn wait_for_start(&self) {
-        wait_until!(
-            CONNECTION_TIMEOUT,
-            if let Ok(mut stream) = tokio::net::TcpStream::connect(self.addr()).await {
-                stream.shutdown().await.unwrap();
-                true
-            } else {
-                false
-            }
-        );
+        Node {
+            child,
+            addr: self.conf.local_addr,
+        }
     }
+}
 
-    fn generate_config_file(&self) -> Result<()> {
-        let path = self.config.path.join(RIPPLED_CONFIG);
-        let content = RippledConfigFile::generate(&self.config)?;
-        fs::write(path, content)?;
+/// Fields to be written to the node's configuration file.
+#[derive(Debug)]
+pub struct NodeConf {
+    /// The path of the cache directory of the node.
+    pub path: PathBuf,
+    /// The socket address of the node.
+    pub local_addr: SocketAddr,
+    /// The initial peer set of the node.
+    pub initial_peers: HashSet<SocketAddr>,
+    /// The initial max number of peer connections to allow.
+    pub max_peers: usize,
+    /// Token when run as a validator.
+    pub validator_token: Option<String>,
+    /// Network's id to form an isolated testnet.
+    pub network_id: Option<u32>,
+}
 
-        Ok(())
+impl NodeConf {
+    pub fn new(path: PathBuf) -> Self {
+        Self {
+            path,
+            local_addr: SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, DEFAULT_PORT)),
+            initial_peers: Default::default(),
+            max_peers: 0,
+            validator_token: None,
+            network_id: None,
+        }
     }
+}
 
-    fn cleanup(&self) -> io::Result<()> {
-        self.cleanup_config_file()?;
-        self.cleanup_cache()
-    }
+pub struct Node {
```

Comment:
The node should hold the `NodeConfig` and the `NodeMetaData` since we may need access to those fields in future test assertions.

Reply:
Done